cmd/metalctl/app/move.go

```diff
@@ -0,0 +1,347 @@
+}
+
+func clearFields(obj client.Object) map[string]interface{} {
+	so, _ := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+
+	for _, field := range []string{"creationTimestamp", "resourceVersion", "uid", "generation", "managedFields"} {
+		delete(so["metadata"].(map[string]interface{}), field)
+	}
+
+	if so["status"] != nil && so["status"].(map[string]interface{})["conditions"] != nil {
+		for _, field := range so["status"].(map[string]interface{})["conditions"].([]interface{}) {
+			delete(field.(map[string]interface{}), "lastTransitionTime")
+		}
+	}
+
+	return so
+}
+
+func getCrdsToBeMoved(ctx context.Context, targetClient client.Client, sourceCrds []*apiextensionsv1.CustomResourceDefinition) ([]*apiextensionsv1.CustomResourceDefinition, error) {
+	crdsToMove := make([]*apiextensionsv1.CustomResourceDefinition, 0, len(sourceCrds))
+	for _, sourceCrd := range sourceCrds {
+		targetCrd := sourceCrd.DeepCopy()
+		err := targetClient.Get(ctx, client.ObjectKeyFromObject(sourceCrd), targetCrd)
+		if apierrors.IsNotFound(err) {
+			crdsToMove = append(crdsToMove, sourceCrd)
+			continue
+		}
+
+		if err != nil {
+			return nil, fmt.Errorf("failed to check CRD existence in the target cluster: %w", err)
+		}
+
+		if reflect.DeepEqual(clearFields(sourceCrd), clearFields(targetCrd)) {
+			slog.Debug("source and target CRDs are the same", slog.String("CRD", crdKind(sourceCrd)))
+			continue
+		}
+		return nil, fmt.Errorf("a CRD %s exists in the target cluster and is different then in the source cluster", sourceCrd.GetName())
+	}
+	return crdsToMove, nil
+}
+
+func getCrsToBeMoved(ctx context.Context, targetClient client.Client, sourceCrs []*unstructured.Unstructured) ([]*unstructured.Unstructured, error) {
+	crsToMove := make([]*unstructured.Unstructured, 0, len(sourceCrs))
+	for _, sourceCr := range sourceCrs {
+		targetCr := sourceCr.DeepCopy()
+		err := targetClient.Get(ctx, client.ObjectKeyFromObject(sourceCr), targetCr)
+		if apierrors.IsNotFound(err) || meta.IsNoMatchError(err) {
+			crsToMove = append(crsToMove, sourceCr)
+			continue
+		}
+
+		if err != nil {
+			return nil, fmt.Errorf("failed to check CR existence in the target cluster: %w", err)
+		}
+
+		if reflect.DeepEqual(clearFields(sourceCr), clearFields(targetCr)) {
+			slog.Debug("source and target CRs are the same", slog.String("CR", crName(sourceCr)))
+			continue
+		}
+		return nil, fmt.Errorf("a CR %s/%s already exists in the target cluster and is different then in the source cluster", sourceCr.GetNamespace(), sourceCr.GetName())
+	}
+	return crsToMove, nil
+}
+
+type Node struct {
+	Cr       *unstructured.Unstructured
+	Children []*Node
+}
+
+func crsOwnerReferenceTrees(crs []*unstructured.Unstructured) []*Node {
+	nodeMap := make(map[types.UID]*Node)
+
+	for _, cr := range crs {
+		nodeMap[cr.GetUID()] = &Node{Cr: cr}
+	}
+	roots := []*Node{}
+	for _, cr := range crs {
+		if len(cr.GetOwnerReferences()) == 0 || nodeMap[cr.GetOwnerReferences()[0].UID] == nil {
```

Comment:
Is the `nodeMap[cr.GetOwnerReferences()[0].UID] == nil` for the case that an owner exists on a CR, but is not a metal related resource?